components/scream/src/share/grid/remap/horizontal_remap_utility.hpp

```diff
@@ -0,0 +1,163 @@
+#ifndef EAMXX_HORIZONTAL_REMAP_UTILITY_HPP
+#define EAMXX_HORIZONTAL_REMAP_UTILITY_HPP
+
+#include "share/grid/abstract_grid.hpp"
+#include "share/io/scorpio_input.hpp"
+#include "share/scream_types.hpp"
+#include "ekat/kokkos/ekat_subview_utils.hpp"
+
+#include <numeric>
+
+namespace scream {
+
+/*===============================================================================================*/
+/*                                        GSSegment
+ * Lightwieght structure to represent a single remapping of source data to a single column.
+ *     Y_target = sum_(n=1)^N ( w_n * Y_source_n )
+ * See description of GSMap structure for more details on the mapping.
+ *
+ * This structure is used to organize the overall horizontal remapping stored in the GSMap structure.
+ * Each segment represents a single target column in the GSMap.
+ * --------------------------------------
+ *  A.S. Donahue (LLNL): 2022-09-07
+ *===============================================================================================*/
+struct GSSegment {
+
+  using gid_type = AbstractGrid::gid_type;
+  using KT = KokkosTypes<DefaultDevice>;
+
+  template <typename S>
+  using view_1d = typename KT::template view_1d<S>;
+  
+  template <typename S>
+  using view_2d = typename KT::template view_2d<S>;
+  
+  template <typename S>
+  using view_3d = typename KT::template view_3d<S>;
+  
+  // Constructors/Destructor
+  GSSegment() {};
+  GSSegment(const gid_type dof_gid, const Int length);
+  GSSegment(const gid_type dof_gid, const Int length, const view_1d<const gid_type>& source_dofs, const view_1d<const Real>& weights);
+
+  // Applying a remap segment
+  void apply_segment(const view_1d<const Real>& source_data, Real& remapped_value);
+  void apply_segment(const Int n1, const view_2d<const Real>& source_data, const view_1d<Real>& remapped_data);
+  void apply_segment(const Int n1, const Int n2, const view_3d<const Real>& source_data, const view_2d<Real>& remapped_data);
+
+  // Helper Functions
+  bool check() const;      // Check if this segment is valid
+  void print_seg() const;  // Useful for debugging, print the segment mapping info
+
+  // Setter Functions
+  void set_dof_idx(const int idx) { m_dof_idx = idx; }
+
+  // Getter Functions
+  gid_type get_dof()     const { return m_dof; }
+  Int      get_dof_idx() const { return m_dof_idx; }
+  Int      get_length()  const { return m_length; }
+  view_1d<gid_type> get_source_dofs() const { return m_source_dofs; }
+  view_1d<Int>      get_source_idx()  const { return m_source_idx; }
+  view_1d<Real>     get_weights()     const { return m_weights; }
+ 
+  // TODO: Not sure why, but this can't be set as private, otherwise `set_dof_idx` doesn't work. 
+  Int      m_dof_idx = -999; // The degree of freedom w.r.t. to the local index for this map
```

Comment:
Why are you initializing this variable here and not in the constructor?

Reply:
Also, is there a different between `int` and `Int`? If not, let's ditch `Int`.

Reply:
Re `int` vs `Int`. I think the odds we'll ever switch the typedef to something different from `int` are miniscule. I always use `int`, but that's my preference. OTOH, I'm in favor of using `intXYZ_t` (X=64,32.16...) when the size is important.

Reply:
As for init-ing at declaration time (inside the class), I like it, actually. It is especially useful when the value is something true, regardless of which ctor is called (I use it a lot for stuff like `m_is_inited`, which should always be false upon construction).

More info: https://isocpp.org/wiki/faq/cpp11-language-classes#member-init

Reply:
Yes, my motivation for init'ing like this is because this is the init value regardless of constructor.  As for `int` vs. `Int`.  I'm pretty agnostic on the syntax.  I can switch everything to `int`.  I recall we tried sticking with `Int` as a declared SCREAM type near the beginning of the project, like 'Real', and I've tried to stick with that convention ever since.